packages/solid-query/src/createBaseQuery.ts

```diff
@@ -45,40 +50,109 @@ export function createBaseQuery<
     observer.getOptimisticResult(defaultedOptions),
   )
 
-  const [dataResource, { refetch, mutate }] = createResource<TData | undefined>(
+  const createServerSubscriber = (
+    resolve: (
+      data:
+        | QueryObserverResult<TData, TError>
+        | PromiseLike<QueryObserverResult<TData, TError> | undefined>
+        | undefined,
+    ) => void,
+  ) => {
+    return observer.subscribe((result) => {
+      notifyManager.batchCalls(() => {
+        const unwrappedResult = { ...unwrap(result) }
+
+        setState(unwrappedResult)
+        resolve(unwrappedResult)
+      })()
+    })
+  }
+
+  const createClientSubscriber = (
+    mutate: () => Setter<QueryObserverResult<TData, TError> | undefined>,
+    refetch: () => void,
+  ) => {
+    return observer.subscribe((result) => {
+      notifyManager.batchCalls(() => {
+        const unwrappedResult = { ...unwrap(result) }
+
+        batch(() => {
+          setState(unwrappedResult)
+          mutate()(() => unwrappedResult)
+          refetch()
+        })
+      })()
+    })
+  }
+
+  /**
+   * Unsubscribe is set lazily, so that we can subscribe after hydration when needed.
+   */
+  let unsubscribe: (() => void) | null = null
+
+  const [queryResource, { refetch, mutate }] = createResource<
+    QueryObserverResult<TData, TError> | undefined
+  >(
     () => {
       return new Promise((resolve) => {
-        if (!(state.isFetching && state.isLoading)) {
-          if ((unwrap(state.data) as TData | typeof emptyData) === emptyData) {
-            resolve(undefined)
+        if (isServer) {
+          unsubscribe = createServerSubscriber(resolve)
+        } else {
+          if (!unsubscribe) {
+            unsubscribe = createClientSubscriber(
+              () => mutate,
+              () => refetch(),
+            )
+          }
+
+          if (!state.isInitialLoading) {
+            resolve(state)
           }
-          resolve(unwrap(state.data))
         }
       })
     },
-  )
+    {
+      get deferStream() {
+        return options().deferStream
+      },
 
-  batch(() => {
-    mutate(() => unwrap(state.data))
-    refetch()
-  })
+      /**
+       * If this resource was populated on the server (either sync render, or streamed in over time), onHydrated
+       * will be called. This is the point at which we can hydrate the query cache state, and setup the query subscriber.
+       *
+       * Leveraging onHydrated allows us to plug into the async and streaming support that solidjs resources already support.
+       *
+       * Note that this is only invoked on the client, for queries that were originally run on the server.
+       */
+      onHydrated(_k, info) {
+        if (info.value) {
+          hydrate(queryClient(), {
+            queries: [
+              {
+                queryKey: defaultedOptions.queryKey,
+                queryHash: defaultedOptions.queryHash,
+                state: info.value,
+              },
+            ],
+          })
```

Comment:
Here I'm assuming that hydrating mutations from server isn't really a use case - or is it?